hawc/apps/materialized/managers.py

```diff
@@ -1,10 +1,151 @@
+from typing import List, Optional
+
+import pandas as pd
+from django.apps import apps
+from django.contrib.contenttypes.models import ContentType
 from django.db import models
 
 
 class FinalRiskOfBiasScoreQuerySet(models.QuerySet):
-    pass
+    def _default_tuples(self, score_values, study_id):
+        return [
+            (score["metric_id"], score)
+            for score in score_values
+            if study_id == score["study_id"] and score["is_default"]
+        ]
+
+    def _override_tuples(self, score_values, study_id, override_model, object_id):
+        content_type_id = ContentType.objects.get_for_model(override_model).id
+        return [
+            (score["metric_id"], score)
+            for score in score_values
+            if study_id == score["study_id"]
+            and score["content_type_id"] == content_type_id
+            and object_id == score["object_id"]
+        ]
+
+    def endpoint_scores(self, endpoint_ids: List[int]):
+
+        Endpoint = apps.get_model("animal", "Endpoint")
+        AnimalGroup = apps.get_model("animal", "AnimalGroup")
+
+        endpoints = Endpoint.objects.filter(pk__in=endpoint_ids).select_related(
+            "animal_group__experiment"
+        )
+
+        score_values = self.values()
+        endpoint_scores = {endpoint_id: {} for endpoint_id in endpoint_ids}
+
+        def default_scores(endpoint, score_values):
+            return [
+                (score["metric_id"], score)
+                for score in score_values
+                if endpoint.animal_group.experiment.study_id == score["study_id"]
+                and score["is_default"]
+            ]
+
+        def animal_group_overrides(endpoint, score_values):
+            content_type_id = ContentType.objects.get_for_model(AnimalGroup).id
+            return [
+                (score["metric_id"], score)
+                for score in score_values
+                if endpoint.animal_group.experiment.study_id == score["study_id"]
+                and score["content_type_id"] == content_type_id
+                and endpoint.animal_group_id == score["object_id"]
+            ]
+
+        def endpoint_overrides(endpoint, score_values):
+            content_type_id = ContentType.objects.get_for_model(Endpoint).id
+            return [
+                (score["metric_id"], score)
+                for score in score_values
+                if endpoint.animal_group.experiment.study_id == score["study_id"]
+                and score["content_type_id"] == content_type_id
+                and endpoint.id == score["object_id"]
+            ]
+
+        for endpoint in endpoints:
+            for metric_id, score in default_scores(endpoint, score_values):
```

Comment:
could be a good use of itertools.chain

https://docs.python.org/3/library/itertools.html#itertools.chain